eip712-adapter-moshi/src/main/java/pm/gnosis/eip712/adapters/moshi/MoshiAdapter.kt

```diff
@@ -0,0 +1,61 @@
+package pm.gnosis.eip712.adapters.moshi
+
+import com.squareup.moshi.Json
+import com.squareup.moshi.JsonClass
+import com.squareup.moshi.Moshi
+import com.squareup.moshi.Types
+import okio.buffer
+import okio.source
+import pm.gnosis.eip712.JsonAdapter
+import pm.gnosis.utils.asEthereumAddress
+import pm.gnosis.utils.nullOnThrow
+import java.io.InputStream
+
+class MoshiAdapter : JsonAdapter {
+    private val moshi by lazy { Moshi.Builder().build() }
+
+    override fun parse(rawJson: String): JsonAdapter.Result {
+        val type = Types.newParameterizedType(List::class.java, Any::class.java)
+        val root = moshi.adapter<List<Any>>(type).fromJson(rawJson) ?: throw IllegalArgumentException("Root is not a Json Array")
+        return parse(root)
+    }
+
+    override fun parse(inputStream: InputStream): JsonAdapter.Result =
+        inputStream.source().buffer().use { bufferedSource ->
+            val type = Types.newParameterizedType(List::class.java, Any::class.java)
+            val root = moshi.adapter<List<Any>>(type).fromJson(bufferedSource) ?: throw IllegalArgumentException("Root is not a Json Array")
+            return parse(root)
+        }
+
+    private fun parse(root: List<Any>): JsonAdapter.Result {
+        val signer = nullOnThrow { root[0].toString().asEthereumAddress() }
+            ?: throw IllegalArgumentException("First element is not a valid signer")
+        val typedDataAsMap = root[1] as? Map<*, *>
+            ?: throw IllegalArgumentException("Second element is not an object")
+        // Hack: converting back to json string (not the most efficient) - see https://github.com/square/moshi/issues/318
+        val rawTypedData = moshi.adapter(Map::class.java).toJson(typedDataAsMap)
+        val typedData = nullOnThrow { moshi.adapter(TypedData::class.java).fromJson(rawTypedData) }
+            ?: throw IllegalArgumentException("Json spec is not TypedData")
+        return JsonAdapter.Result(
+            signer = signer,
+            primaryType = typedData.primaryType,
+            domain = typedData.domain,
+            message = typedData.message,
+            types = typedData.types.mapValues { (_, types) -> types.map { JsonAdapter.TypeParam(it.name, it.type) } }
+        )
+    }
+
+    @JsonClass(generateAdapter = true)
+    data class TypedData(
+        @Json(name = "types") val types: Map<String, List<TypeParam>>,
+        @Json(name = "primaryType") val primaryType: String,
+        @Json(name = "domain") val domain: Map<String, Any>,
```

Comment:
Shouldn't this be `List<TypeParam>`

Reply:
I thought about it but if domain in the future is extended to have objects then `List<TypeParam>` would not work. I didn't assume it'd be a map between `parameterName` and `parameterValue`

Reply:
Don't plan for the future ;) at least not that far :P 

I don't think the domain structure will change anytime soon

Reply:
Issue with this is that `EIP712JsonParser.buildStruct712` expects `values: Map<String, Any>`